src/atomEffect.ts

```diff
@@ -1,30 +1,40 @@
-import type { Getter, Setter } from 'jotai/vanilla'
+import type { Atom, Getter, Setter } from 'jotai/vanilla'
 import { atom } from 'jotai/vanilla'
 
 type CleanupFn = () => void
 
+type SetterWithRecurse = Setter & { recurse: Setter }
+
 export function atomEffect(
-  effectFn: (get: Getter, set: Setter) => void | CleanupFn
+  effectFn: (get: Getter, set: SetterWithRecurse) => void | CleanupFn
 ) {
   const refAtom = atom(() => ({
     mounted: false,
     inProgress: 0,
     promise: undefined as Promise<void> | undefined,
     cleanup: undefined as CleanupFn | void,
+    fromCleanup: false,
+    recursing: false,
+    refresh: () => {},
+    refreshing: false,
+    set: (() => {}) as Setter,
   }))
-  if (process.env.NODE_ENV !== 'production') {
-    refAtom.debugPrivate = true
-  }
 
   const refreshAtom = atom(0)
-  if (process.env.NODE_ENV !== 'production') {
-    refreshAtom.debugPrivate = true
-  }
 
   const initAtom = atom(null, (get, set, mounted: boolean) => {
     const ref = get(refAtom)
     ref.mounted = mounted
     if (mounted) {
+      ref.set = set
```

Comment:
There was simply no other way. After the first recursion, the effectFn and setter must be run synchronous so that it can recurse down the call stack. So we have to use set instead of setSelf.